### PR TITLE
[AIROCMLIR-707] Fix split-kv attention masking and sweep RMS for attention configs

### PR DESCRIPTION
**Pull request by bogdan-petkovic**
## Motivation

Attention performance sweeps were failing on split-KV configurations (`split_kv > 1`) with RMS validation errors, NaNs, OOM crashes, or invalid results. The failures showed up across causal masking, GQA, KV-cache style `current_seqlen`, and mixed dtypes (`f16`, `bf16`, `i8`), including cases with `trans_q` and bias.

The goal is to make split-KV attention sweeps reliable without weakening the default kernel verifier policy. Fixes target:
- GPU split-KV softmax updates that produced NaNs on fully masked or empty key partitions.
- GPU split-KV iteration math that yielded zero iterations per split (and downstream 0/0 in `scaleFinalOutput`) for small `seq_len_k` with large `split_kv`.
- Host reference masking and combine logic in `rocmlir-gen` that did not match split-KV behavior under causal masking.
- Sweep validation tolerance so attention configs are checked against a realistic numeric band instead of failing on reference-path noise.
- Sweep generator behavior so memory-heavy split-KV cases don't surface in CI as OOM-crashed sweep samples.

`rocmlir-gen`'s default `-RMS_threshold` for fp16/bf16 without an explicit override stays `0.001`. Only the attention sweep driver gets a separate default when the config does not set a threshold.

## Technical Details

**Kernel path: guard `-inf - (-inf)` in split-KV softmax updates**
In `mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp`, split-KV attention rewrites update row-wise softmax state across KV partitions. When a partition has no valid keys, both the running row max and the partition max can be `-inf`, so `exp2(score - max)` and `exp2(old_max - new_max)` can see `-inf - (-inf)` and become NaN. That poisons row sums and downstream combine.

The change detects the case where both operands are `-inf` and uses `0` before `exp2`, so empty or fully masked partitions contribute zero instead of NaN. This is a correctness fix on the kernel path, not a verifier relaxation.

**Kernel path: non-causal split-KV iteration math + `scaleFinalOutput` 0/0 guard**
The non-causal / non-KV-cache split-KV branch in `GridwiseAttentionAccelRewritePattern` computed per-split iterations as `gemm0M / (gemm0MPerBlock * splitKV)` using truncating integer division. When `gemm0MBlocks < splitKV` (small `seq_len_k` with large `split_kv`), this evaluates to `0` for every split, so every split-block skips the softmax loop entirely. The kernel then divides the (zero) output by `sum = 0` in `scaleFinalOutput`, producing NaNs that propagate through the host combine stage.

This is fixed by:
- Using ceil-division to compute iterations per split, then clamping `end` to `gemm0MBlocks` so trailing splits where `start >= gemm0MBlocks` become cleanly empty.
- A defensive `0/0` guard in `scaleFinalOutput`: when a row's sum is exactly zero, the per-split output stores `0` instead of `NaN`. The host combine stage already tolerates `-inf` max, but only when each per-split partial output is finite.

**Host reference: causal valid-split masking**
In `mlir/tools/rocmlir-gen/rocmlir-gen.cpp`, `computeValidSplitKV()` used a causal branch that forced `currSeqLen = 0`, which made every split invalid and did not reflect per-query causal reach. Causal masking now builds a per-(batch-head, query-row) valid split count from each row's effective key length (including `prefix_offset` when present). Non-causal configs keep the per-batch-head path. The `usePerRowMask` predicate also triggers for non-empty `prefix_offset`, mirroring how the kernel treats prefix-causal as causal.

`createMaskSplitKV()` accepts either per-batch-head or per-(batch-head, query-row) `validSplitKV` layouts and broadcasts the threshold tensor accordingly. `computeFinalAttentionStage()` still masks invalid splits on both the partial output and LSE tensors before the split-KV combine; its size assertion now accepts either layout.

**Host reference: f32 combine for narrow floats**
For `f16` and `bf16`, the split-KV combine stage (reduce-max, exp, weighted sum, normalization) runs in `f32` and casts the final combined result back to storage type. That reduces accumulation error in the reference path when comparing against the GPU kernel during sweeps.

**Sweep policy: default attention RMS 0.005**
In `mlir/utils/performance/parameterSweeps.py`, attention sweeps without an explicit `-RMS_threshold` now append `-RMS_threshold 0.005` for all attention dtypes (`bf16` keeps `0.01`). The band covers observed sweep disagreement (including i8 with `trans_q` and high `split_kv`) while staying tighter than the old bf16-only default.

Configs that set their own `-RMS_threshold` are unchanged.

**Sweep policy: widen RMS band for unscaled large-head_dim attention**
Sampled configs with `with_attn_scale=False` and `head_dim_qk > 64` saturate softmax (`|QK| ~ O(sqrt(d))` collapses to near one-hot), so CPU vs GPU float-arithmetic ordering inside `exp`/accumulate dominates the diff (observed RMS up to ~6% in bf16, ~1% in f16/i8) independent of `split_kv`. For this regime only, `test_config` widens `-RMS_threshold` to `0.15` so the verifier still catches NaN/crash regressions but no longer false-fails on known float associativity in saturated softmax. All other configs are unaffected.

**Sweep generator: device-memory-aware split-KV prefilter** (folds in draft #2366)
In `mlir/utils/performance/attentionSweeps.py`, the sweep generator now estimates extra split-KV temporary storage for each sampled shape and rejects samples above a budget *before* generating MLIR. The budget defaults to `deviceMem / 8`, clamped to `[1 GiB, 8 GiB]`, with a `1.5 GiB` fallback if the HIP query fails. A `--splitkv-extra-bytes-limit` CLI override is available. Filter-out reasons are now tracked separately (`MAX_TOKENS` vs `splitKV extra-storage`) and reported cumulatively across initial and refill batches.

No compiler/verifier logic changes for this part; scope is limited to sweep sampling / filtering behavior.

## Test Plan

- Added targeted deterministic tests (lit + E2E) for each kernel/host fix; see **Test Coverage** below. PR CI runs the full build, these new lit tests, the new `PrAttention{F16,BF16,F32}` E2E configs (GPU), Python format/lint, the Python performance script tests, and the `AttentionSweeps` job.
- Locally re-ran the original 8 failing May-3 sweep configs plus 5 newer May-14 sweep failures on `gfx1201` (16 GiB) with all changes in place. 6/13 PASS on the kernel path. Of the remaining 7: 6 are correctly filtered out by the new split-KV prefilter (they were OOM-bound on a 16 GiB GPU because `split_kv = 128` × large `g × num_heads_q × head_dim_v` blew past VRAM), and 1 (split_kv=1, bf16 trans-all) trips only the per-element `relDiff` check on near-zero outputs while passing both RMS and abs-diff — out of scope here and tracked separately.

## Test Coverage

New and updated tests pin each fix above:

**Lit — kernel path (IR lowering)**
- `mlir/test/Dialect/Rock/gridwise_attention_accel_lowering.mlir` and
  `toblockwise_attention_accel_lowering.mlir` (updated): add explicit
  `CHECK`s for the `-inf - (-inf)` softmax guard and the `scaleFinalOutput`
  `0/0` guard, so the guarded select is verified in the emitted IR instead
  of the old unguarded store.
- `mlir/test/Dialect/Rock/gridwise_attention_splitkv_noncausal_lowering.mlir`
  (new): pins the non-causal split-KV ceil-division + `end` clamp iteration
  math, covering the `gemm0MBlocks < splitKV` case.

**Lit — host reference (`rocmlir-gen`)**
- `mlir/test/rocmlir-gen/attention-splitkv-host-perrow-mask.mlir` (new):
  pins the per-(batch-head, query-row) valid-split mask under causal +
  `prefix_offset`.
- `mlir/test/rocmlir-gen/attention-splitkv-host-f32-combine.mlir` (new):
  pins the f16/bf16 → f32 combine and cast-back in the reference path.

**E2E — GPU numeric (run in PR CI)**
- `mlir/test/e2e/PrAttention{F16,BF16,F32}.toml` (new configs): non-causal
  split-KV (basic, trailing/padded splits, GQA) and causal split-KV with
  `seq_len_q > 1` (per-row mask, GQA, `prefix_offset`), exercising the
  guards and combine on real hardware.

The sweep-policy (`RMS` band) and split-KV memory prefilter changes remain
covered by the existing `mlir/utils/performance` pytest suite and the
`AttentionSweeps` job.

## Test Result

- [x] PR CI
- [x] Nightly CI
- [ ] Weekly CI

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 1]
```diff
@@ -870,13 +896,17 @@ struct GridwiseAttentionAccelRewritePattern
           rewriter, loc, ldMaxRowBuffer, ldgemm0OutBufferMax);
 
       // ldGemm0OutSubMaxExp = exp(gemm0Out  -maxRowBufferNew)
-      Type ldGemm0OutElemType = getElementTypeOrSelf(gemm0Out.getType());
-      Value ldGemm0Out = InBoundsLoadOp::create(
-          rewriter, loc, ldGemm0OutElemType, gemm0Out, gemm0OutCoords);
+      Value ldGemm0Out = InBoundsLoadOp::create(rewriter, loc, gemm0OutElemType,
```
Comment (umangyadav) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:899]: can you add some lit tests and E2E tests that would exercise this path ?

**Review by umangyadav — APPROVED** [review]
Would be good to have a seperate PRs for each bug/fixes. Okay to land this PR as it is
